File: python/riot_sdk/core/ready_for_next_order.py
from __future__ import annotations

SUCCESS_ORDER_STATE = 5
IDLE_PROC_STATE = "IDLE"


def is_ready_for_next_order(
    order_state: int,
    proc_state: str | None,
    processing_order: bool = False,
) -> bool:
    """Pure ReadyForNextOrder check (BC-STATE-003 / ADR-0008).

    True only when previous order is SUCCESS and vehicle is IDLE.
    AWAITING_ORDER alone is not sufficient.
    """
    return (
        order_state == SUCCESS_ORDER_STATE
        and proc_state == IDLE_PROC_STATE
        and not processing_order
    )

File: csharp/RIoT.Sdk.Core/ReadyForNextOrder.cs
namespace RIoT.Sdk.Core;

/// <summary>
/// Pure ReadyForNextOrder check (BC-STATE-003 / ADR-0008).
/// Safe to dispatch next order only when previous order is SUCCESS and vehicle is IDLE.
/// Not ArrivalAtStation; AWAITING_ORDER alone is not sufficient.
/// </summary>
public static class ReadyForNextOrder
{
    /// <summary>RIoT orderState for SUCCESS.</summary>
    public const int SuccessOrderState = 5;

    /// <summary>RIoT vehicle procState for IDLE.</summary>
    public const string IdleProcState = "IDLE";

    public static bool IsReady(int orderState, string? procState, bool processingOrder = false)
        => orderState == SuccessOrderState
           && string.Equals(procState, IdleProcState, StringComparison.Ordinal)
           && !processingOrder;
}
